packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+    });
+
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackages");
+        // Step 2: Rewrite packages. Can be done in parallel (and w/ step 3) if this becomes a bottleneck.
+        for (auto &file : files) {
+            if (file.file.data(gs).sourceType == core::File::Type::Package) {
+                auto &packageInfo = packageInfoByFile[file.file];
+                if (packageInfo != nullptr) {
+                    core::Context ctx(gs, core::Symbols::root(), file.file);
+                    file = rewritePackage(ctx, move(file), *packageInfo, packageInfoByMangledName);
+                }
+            }
+        }
+    }
+
+    // Step 3: Find files within each package and rewrite each.
+    {
+        Timer timeit(gs.tracer(), "packager.rewritePackagedFiles");
+
+        auto resultq = make_shared<BlockingBoundedQueue<vector<ast::ParsedFile>>>(files.size());
+        auto fileq = make_shared<ConcurrentBoundedQueue<ast::ParsedFile>>(files.size());
+        for (auto &file : files) {
+            fileq->push(move(file), 1);
+        }
+
+        workers.multiplexJob("rewritePackagedFiles", [&gs, &packages, fileq, resultq]() {
+            Timer timeit(gs.tracer(), "packager.rewritePackagedFilesWorker");
+            vector<ast::ParsedFile> results;
+            u4 filesProcessed = 0;
+            ast::ParsedFile job;
+            for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
+                if (result.gotItem()) {
+                    filesProcessed++;
+                    if (job.file.data(gs).sourceType == core::File::Type::Normal) {
+                        // Check if file path is part of a package.
+                        // TODO(jvilk): Could use a prefix array to make this lookup more efficient.
+                        auto path = job.file.data(gs).path();
+                        bool packaged = false;
+                        core::Context ctx(gs, core::Symbols::root(), job.file);
+                        for (const auto &pkg : packages) {
+                            if (absl::StartsWith(path, pkg->packagePathPrefix)) {
+                                job = rewritePackagedFile(ctx, move(job), pkg->name.mangledName);
+                                packaged = true;
+                                break;
+                            }
+                        }
+                        if (!packaged) {
+                            // Don't transform, but raise an error.
+                            if (auto e = ctx.beginError(job.tree->loc, core::errors::Packager::UnpackagedFile)) {
+                                e.setHeader("File `{}` does not belong to a package; add a `__package.rb` file to one "
+                                            "of its parent directories",
+                                            ctx.file.data(gs).path());
+                            }
+                        }
+                    }
```

Comment:
iiuc, this makes it such that every file must belong to a package, and it won't be allowed to have unpackaged files? Or is that a misunderstanding of this code?

Reply:
Yup. There are no unpackaged files.

In practice, this means Stripe will have a top-level package named "Unpackaged" that contains all code, and subpackages will be chipped off from that.